sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigtable/BigtableServiceImpl.java

```diff
@@ -210,6 +234,246 @@ public Row getCurrentRow() throws NoSuchElementException {
     }
   }
 
+  @VisibleForTesting
+  static class BigtableMiniBatchReaderImpl implements Reader {
+    private BigtableSession session;
+    private final BigtableSource source;
+    private Row currentRow;
+    private Queue<FlatRow> buffer;
+    private RowSet rowSet;
+    private ServiceCallMetric serviceCallMetric;
+    private Future<List<FlatRow>> future;
+    private ByteString lastFetchedRow;
+    private boolean lastFillComplete;
+
+    private int miniBatchLimit = DEFAULT_MINI_BATCH_SIZE;
+    private long bufferSizeLimit = DEFAULT_BYTE_LIMIT;
+    private int miniBatchWaterMark;
+    private final String tableName;
+
+    @VisibleForTesting
+    BigtableMiniBatchReaderImpl(BigtableSession session, BigtableSource source) {
+      this.session = session;
+      if (source.getMaxBufferElementCount() != null && source.getMaxBufferElementCount() != 0) {
+        this.miniBatchLimit = source.getMaxBufferElementCount();
+        this.bufferSizeLimit = (long) miniBatchLimit * 100 * 1024 * 1024;
```

Comment:
Please extract default consts to named consts. ie private final static long DEFAULT_BUFFER_SIZE_LIMIT = ...

Also the bufferSizeLimit should be a function of memory size with hard limits for the max & min. See the `Runtime` class for ways to inspect the available memory resources